### PR TITLE
feat(aws-lambda): add new configuration field `empty_arrays_mode` to control empty array decoding behavior

### PR DESCRIPTION
**Pull request by windmgc**
### Summary

This PR adds a new configuration field `empty_arrays_mode` to control the behaviour that whether Kong will transform an empty array `[]`(returned by Lambda function) into an empty object `{}` when enabled aws-lambda plugin in the response.

**Essentially** this PR is just a small bugfix for not previously enabling `decode_array_with_array_mt` when decoding Lambda function's response. However, enabling `decode_array_with_array_mt` may bring unexpected behaviour changes on the user side(since the wrong behaviour itself has existed for long), so we decided to add a new field to keep backward compatibility, and remove this field in the next major version to let the correct behaviour(keep empty arrrays as is) to become the default behaviour in the next major version.

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

FTI-5937

### CODE REVIEW

File: spec/03-plugins/27-aws-lambda/99-access_spec.lua [hunk 1]
```diff
@@ -923,6 +961,30 @@ for _, strategy in helpers.each_strategy() do
         end, 10)
       end)
 
+      it("#testme invokes a Lambda function with empty array", function()
```
Comment (vm-001) [inline spec/03-plugins/27-aws-lambda/99-access_spec.lua:964]:
```suggestion
      it("invokes a Lambda function with empty array", function()
```

**Review by vm-001 — CHANGES_REQUESTED** [review]

File: changelog/unreleased/kong/feat-aws-lambda-decode-empty-array.yml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+message: |
+    **AWS-Lambda**: A new configuration field `empty_arrays_mode` is now added to control whether Kong should send `[]` empty arrays(returned by Lambda function) as `[]` empty arrays or `{}` empty objects in JSON responses.`
```
Comment (outsinre) [inline changelog/unreleased/kong/feat-aws-lambda-decode-empty-array.yml:2]:
```suggestion
    **AWS-Lambda**: A new configuration field `empty_arrays_mode` is now added to control whether Kong should send `[]` empty arrays (returned by Lambda function) as `[]` empty arrays or `{}` empty objects in JSON responses.`
```

File: kong/plugins/aws-lambda/schema.lua [hunk 3]
```diff
@@ -116,6 +116,13 @@ return {
           default = "v1",
           one_of = { "v1", "v2" }
         } },
+        { empty_arrays_mode = { -- TODO: this config field is added for backward compatibility and will be removed in next major version
+          description = "An optional value that defines whether Kong should send empty arrays(returned by Lambda function) as `[]` arrays or `{}` objects in JSON responses. The value `legacy` means Kong will send empty arrays as `{}` objects in response",
```
Comment (outsinre) [inline kong/plugins/aws-lambda/schema.lua:120]: Is this description correct?

```
Kong will send empty arrays as `{}` objects
```

Comment (outsinre) [inline kong/plugins/aws-lambda/schema.lua:120]: How about this?

```suggestion
          description = "An optional value that defines whether Kong should send empty arrays(returned by Lambda function) as `[]` arrays or `{}` objects in JSON responses. The value `legacy` means Kong will send empty arrays `[]` as `{}` objects in response",
```

File: spec/03-plugins/27-aws-lambda/99-access_spec.lua (hunk 1 above)
Comment (windmgc) [reply spec/03-plugins/27-aws-lambda/99-access_spec.lua:964]: nice catch

**Review by vm-001 — APPROVED** [review]
Looks good. 

I would love to see if we have a compatibility test for that new field.

**Review by tzssangglass — APPROVED** [review]

File: kong/plugins/aws-lambda/handler.lua [hunk 4]
```diff
@@ -240,6 +241,16 @@ function AWSLambdaHandler:access(conf)
...
+  -- TODO: remove this in the next major release
+  -- function to remove array_mt metatables from empty tables
+  -- This is just a backward compatibility code to keep a
+  -- long-lived behavior that Kong responsed JSON objects
+  -- instead of JSON arrays for empty arrays.
+  local ct = headers["Content-Type"]
+  if ct and ct:lower():match("application/.*json") and conf.empty_arrays_mode == "legacy" then
+    content = remove_array_mt_for_empty_table(content)
+  end
+
```
Comment (bungle) [inline kong/plugins/aws-lambda/handler.lua:253]: Faster checks first.

```suggestion
  if conf.empty_arrays_mode == "legacy" then
    local ct = headers["Content-Type"]
    if ct and ct:lower():match("application/.*json") and  then
      content = remove_array_mt_for_empty_table(content)
    end
  end
```

**Review by bungle — CHANGES_REQUESTED** [review]
Just a small comment.

File: kong/plugins/aws-lambda/handler.lua (hunk 4 above)
Comment (bungle) [reply kong/plugins/aws-lambda/handler.lua:253]: Where does this `ct:lower():match("application/.*json")` come from, does it come directly from AWS-library?

Comment (windmgc) [reply kong/plugins/aws-lambda/handler.lua]: Yes, it does

Fixed

**Review by bungle — APPROVED** [review]

**Pull request merged**